File: lib/betterbam.rb
require 'ffi'
# require 'betterbam/library'

class BetterBam

  include FFI
  extend FFI::Library

  # ffi_lib Library.load
  ffi_lib "ext/libbetterbam.so"

  attach_function "delete_pointer_to_betterbam_object",
                  "delete_pointer_to_betterbam_object", [:pointer], :void

  attach_function "betterbam_method", "betterbam_method", [:pointer], :int

  attach_function "create_betterbam_object",
                  "create_betterbam_object", [:pointer], :pointer

  attr_accessor :file

  def initialize file
    self.file = file
  end

  def parse_bam
    pointer = FFI::MemoryPointer.from_string(self.file)
    object = create_betterbam_object(pointer)
    puts betterbam_method(object)
    delete_pointer_to_betterbam_object object
  end

end

File: ext/betterbam.c
#include "betterbam_object.hpp"
#include "betterbam.h"


betterbam_object_p create_betterbam_object(char *file) {
  return static_cast<void *>(new betterbam_object(file));
}

void delete_pointer_to_betterbam_object(betterbam_object_p p) {
  delete static_cast<betterbam_object *>(p);
}

int betterbam_method(betterbam_object_p p) {
  betterbam_object o = *static_cast<betterbam_object *>(p);
  return o.betterbam_method();
}
